### ISSUE TITLE
Cannot build on platforms where all sources are filtered out.

### ISSUE THREAD

**Issue by ains** [bazelbuild/rules_go#453]
As of https://github.com/bazelbuild/rules_go/commit/05bf1b2adfbfe4e77552f25e3331e14865bff586 Gazelle will generate BUILD files containing sources that wont be built on the current platform, and leave filtering to the build steps.

However one issue I'm now running into on OSX after rebasing onto master is a external repository where Gazelle generates a target where all source files are filtered out at build time, resulting in an error.

You can reproduce this bug (on OSX) by checking out this tree (https://github.com/ains/netlink/tree/bazel_repro) which is the result of running Gazelle on head of https://github.com/vishvananda/netlink 

>richards-mbp:netlink ains$ bazel build //...
INFO: Found 11 targets...
ERROR: /private/tmp/netlink/BUILD:5:1: error executing shell command: 'bazel-out/local-fastbuild/bin/bazel-out/local-fastbuild/genfiles/cgo_default_library.cgo.dir.CGoCodeGenFile.params' failed: Process exited with status 1 [sandboxed].
no buildable Go source files in //:cgo_default_library.cgo
Use --strategy=CGoCodeGen=standalone to disable sandboxing for the failing actions.
INFO: Elapsed time: 0.836s, Critical Path: 0.56s

The issue is us throwing an error when all source files are filtered out (which is now possible)
https://github.com/bazelbuild/rules_go/blob/master/go/def.bzl#L784

We can't fail silently here (exit 0) as Bazel is expecting output files, maybe we need to hoist the filtering outside of this script and not execute the action if all sources are filtered out?

**Comment by jayconrod** [comment]
This issue is being tracked in #409. There are a couple proposals in there on how to deal with the problem. I'm implementing the [second proposal](https://github.com/bazelbuild/rules_go/issues/409#issuecomment-297532311), but it's a large change, and it will take some time.

TL;DR version: Gazelle will generate rules with `select` expressions for platform-specific sources, dependencies, and options. In your repro example, your dependencies on Linux-specific libraries would be behind the `select`, so they wouldn't be depended on or compiled on Darwin.

Until the full solution is ready, you can restore the old Gazelle behavior by passing the `-build_tags` argument with your GOOS and GOARCH. You can also pass this through `go_repository`.

**Comment by ains** [comment]
@jayconrod thanks for the ticket, looks like a good solution!

**Comment by jayconrod** [comment]
This should be working now that #509 has landed. I just checked out the repo you mentioned and re-ran Gazelle. `bazel build //...` worked.

Let me know if you are still having any issues with this.

**Comment by mterring** [comment]
I've found another case where this can occur - when a `cgo_library` is only applicable for certain platforms.

This can be seen when using https://github.com/rjeczalik/notify on linux as the cgo library is only needed on darwin.

It can be fixed by changing the library to use a select matching the platforms of the `cgo_library`:
```
-    library = ":cgo_default_library",
+    library = select({"@io_bazel_rules_go//go/platform:darwin_amd64": ":cgo_default_library", "//conditions:default": None }),
```
